lib/proxy_api/external_ipam.rb

```diff
@@ -0,0 +1,210 @@
+    #      "00:0a:95:9d:68:10": {"ip": "2001:db8:abcd:12::1", "timestamp": "2019-09-17 12:03:43 -D400"},
+    #      "00:0a:95:9d:68:11": {"ip": "2001:db8:abcd:12::2", "timestamp": "2019-09-17 12:13:22 -D400"}
+    #   }
+    # }
+    #
+    # Params: 1. subnet:  The IPv4 or IPv6 subnet CIDR. (Examples: IPv4 - "100.10.10.0/24", IPv6 - "2001:db8:abcd:12::/124")
+    #         2. mac:     The mac address of the interface obtaining the IP address for
+    #
+    # Returns: A hash with next available IP in the "data" field, or a hash with "error" on failure
+    #
+    # Examples:
+    #   Responses if success:
+    #     IPv4: {"code": 200, "success": true, "data": "100.55.55.3", "time": 0.012}
+    #     IPv6: {"code": 200, "success": true, "data": "2001:db8:abcd:12::1", "time": 0.012}
+    #   Response if missing required params:
+    #     {"error": ["A 'cidr' parameter for the subnet must be provided(e.g. 100.10.10.0/24)","A 'mac' address must be provided(e.g. 00:0a:95:9d:68:10)"]}
+    #   Response if subnet error:
+    #     {"error": "The specified subnet does not exist in External IPAM."}
+    #   Response if there are no free addresses:
+    #     {"error": "No free addresses found"}
+    #   Response if can't connect to External IPAM server
+    #     {"error": "Unable to connect to External IPAM server"}
+    def next_ip(subnet, mac)
+      raise "subnet cannot be nil" if subnet.nil?
+      parse get("/next_ip?cidr=#{subnet}&mac=#{mac}")
+    rescue => e
+      raise ProxyException.new(url, e, N_("Unable to retrieve the next available IP for subnet %s from External IPAM."), subnet)
+    end
+
+    # Adds an IP address to the specified subnet in External IPAM. This will reserve the IP in the
+    # External IPAM database.
+    #
+    # Params: 1. ip:     IP address to be added
+    #         2. subnet: The IPv4 or IPv6 subnet CIDR. (Examples: IPv4 - "100.10.10.0/24", IPv6 - "2001:db8:abcd:12::/124")
+    #
+    # Returns: true if IP was added successfully to External IPAM, otherwise false
+    #
+    # Proxy response Examples:
+    #   Response if success:
+    #     {"code": 201, "success": true, "message": "Address created", "id": "156", "time": 0.015}
+    #   Response if IP already reserved:
+    #     {"code": 409, "success": false, "message": "IP address already exists", "time": 0.006}
+    #   Response if subnet error:
+    #     {"error": "The specified subnet does not exist in External IPAM."}
+    #   Response if missing required params:
+    #     {"error": ["A 'cidr' parameter for the subnet must be provided(e.g. 100.10.10.0/24)","Missing 'ip' parameter. An IPv4 address must be provided(e.g. 100.10.10.22)"]}
+    #   Response if can't connect to External IPAM server
+    #     {"error": "Unable to connect to External IPAM server"}
+    def add_ip_to_subnet(ip, subnet)
+      raise "subnet cannot be nil" if subnet.nil?
+      response = parse post({:ip => ip, :cidr => subnet}, "/add_ip_to_subnet")
```

Comment:
In a REST API `add_ip_to_subnet` sounds odd. You're already sending a `POST` which implies you're adding. That's why you usually have something like `/subnet/#{subnet}` and `POST` the IP there. In that case a HTTP 404 would indicate the subnet doesn't exist. A `GET /subnet/#{subnet}` would list all IPs within that subnet.

Reply:
While I am not big fan of REST and I prefer RPC, Ewoud made a good call here - our proxy API is REST based. Can you redo it to be as RESTy as possible? Basically everything can stay as is, it will be just different URLs and sometimes DELETE/POST instead of GET. That's all.